modules/core/src/test/scala/fs2/kafka/TransactionalKafkaProducerSpec.scala

```diff
@@ -267,4 +269,204 @@ class TransactionalKafkaProducerSpec extends BaseKafkaSpec with OptionValues wit
+
+        val error = new RuntimeException("BOOM")
+
+        val produced =
+          (for {
+            producer <- Stream.resource {
+              builder {
+                producerSettings[IO](config)
+                  .withCreateProducer { properties =>
+                    IO.delay {
+                      new org.apache.kafka.clients.producer.KafkaProducer[Array[Byte], Array[Byte]](
+                        (properties: Map[String, AnyRef]).asJava,
+                        new ByteArraySerializer,
+                        new ByteArraySerializer
+                      ) {
+                        override def sendOffsetsToTransaction(
+                          offsets: util.Map[TopicPartition, OffsetAndMetadata],
+                          consumerGroupId: String
+                        ): Unit =
+                          if (offsets.containsKey(new TopicPartition(topic, 2))) {
+                            throw error
+                          } else {
+                            super.sendOffsetsToTransaction(offsets, consumerGroupId)
+                          }
+                      }
+                    }
+                  }
+              }.withDynamicIds("transactions", noRebalance(topic, 3))
+            }
+            recordsToProduce = toProduce.map {
+              case (key, value) => ProducerRecord(topic, key, value)
+            }
+            offsets = toProduce.mapWithIndex {
+              case (_, i) =>
+                CommittableOffset(
+                  new TopicPartition(topic, i % 3),
+                  new OffsetAndMetadata(i.toLong),
+                  Some("group"),
+                  _ => IO.unit
+                )
+            }
+            records = TransactionalProducerRecords(
+              Chunk.seq(recordsToProduce.zip(offsets)).map {
+                case (record, offset) =>
+                  CommittableProducerRecords(
+                    NonEmptyList.one(record),
+                    offset
+                  )
+              },
+              toPassthrough
+            )
+            result <- Stream.eval(producer.produce(records).attempt)
+          } yield result).compile.lastOrError.unsafeRunSync()
+
+        produced shouldBe Left(error)
+
+        val consumedOrError = {
+          implicit val transactionalConfig: EmbeddedKafkaConfig = EmbeddedKafkaConfig(
+            kafkaPort = config.kafkaPort,
+            zooKeeperPort = config.zooKeeperPort,
+            customConsumerProperties =
+              Map(ConsumerConfig.ISOLATION_LEVEL_CONFIG -> "read_committed")
+          )
+          Either.catchNonFatal(consumeFirstKeyedMessageFrom[String, String](topic))
+        }
+
+        // FIXME: This currently fails because committing offsets only fails in the producer
+        // for one of the three topic/partitions. The other t/p's successfully commit messages,
+        // but we don't currently have a mechanism to signal what worked and what didn't.
```

Comment:
@vlovgr here's a case that I hadn't fully thought through. Now that a `TransactionalKafkaProducer` can wrap multiple `KafkaByteProducer`s, producing a single `TransactionalProducerRecords` isn't atomic any more. As-is, the only way to safely recover from a transaction failure would be to restart the whole stream and re-process any messages that weren't committed. We might be able to do better by raising some custom error containing all the records that _were_ successfully committed, but I'm not sure it's safe / worth it. What do you think?

Reply:
> Now that a `TransactionalKafkaProducer` can wrap multiple `KafkaByteProducer`s, producing a single `TransactionalProducerRecords` isn't atomic any more.

I think users might find it confusing that `TransactionalProducerRecords` is no longer atomic with this construct, but atomic if using a single producer. It might be worth considering an alternative construct or way to communicate what is and isn't atomic. If nothing else, we should try to cleary document this.

> As-is, the only way to safely recover from a transaction failure would be to restart the whole stream and re-process any messages that weren't committed. We might be able to do better by raising some custom error containing all the records that were successfully committed, but I'm not sure it's safe / worth it. What do you think?

Raising the error and letting the user decide sounds acceptable for a first pass. We can explore recovery options (similar to `CommitRecovery`) in a subsequent pull request. Also, the standard producer doesn't recover from any errors currently.

Reply:
One option for dealing with the atomicity problem would be to drop `TransactionalProducerRecords`, and move `passthrough` onto `CommittableProducerRecords`. In https://github.com/ovotech/fs2-kafka/pull/130#discussion_r283091204 it looks like we decided to keep both so users could produce to multiple topic/partitions in a single transaction, but now there isn't really a single transaction so I'm not sure if the abstraction holds up.

Reply:
I wonder if I'm attacking this at the wrong layer. Instead of having a single `TransactionalKafkaProducer` wrap multiple `KafkaByteProducer`s, we could have a 1:1 mapping and then a higher-level helper to manage spinning the `TransactionalKafkaProducer` instances up/down. The helper could be a:
```scala
Pipe[F, CommittableProducerRecords, Chunk[ProducerResult]]
```
Where each emitted `Chunk` originates a single successfully-completed transaction. Batching for transactions could be tweaked using similar arguments as `groupWithin`. If the transaction for a particular topic/partition fails, the pipe would finish emitting the results of any other in-flight transactions, then fail with the error.

What do you think?

Reply:
That sounds like an interesting idea to explore. I agree chunking probably needs to be handled outside, as one needs to manually determine what constitutes a transaction. This also fits well with the existing `produce` pipes.